src/ua_securechannel.c

```diff
@@ -755,15 +755,16 @@ decryptChunk(UA_SecureChannel *channel, const UA_SecurityPolicyCryptoModule *cry
        channel->securityMode == UA_MESSAGESECURITYMODE_SIGNANDENCRYPT ||
        messageType == UA_MESSAGETYPE_OPN) {
         /* Compute the padding size */
-        sigsize = cryptoModule-> getRemoteSignatureSize(securityPolicy, channel->channelContext);
+        sigsize = cryptoModule->getRemoteSignatureSize(securityPolicy, channel->channelContext);
 
         if(channel->securityMode == UA_MESSAGESECURITYMODE_SIGNANDENCRYPT ||
            (messageType == UA_MESSAGETYPE_OPN &&
             channel->securityMode != UA_MESSAGESECURITYMODE_NONE)) {
-            paddingSize = chunk->data[chunkSizeAfterDecryption - sigsize - 1];
+            // we need to add one to the padding size since the paddingSize byte itself need to be removed as well.
+            // TODO: write unit test for correct padding calculation
+            paddingSize = (size_t) chunk->data[chunkSizeAfterDecryption - sigsize - 1] + 1;
```

Comment:
Why would this work for a single chunk but not multiple chunks?
Is the extraPaddingSize considered?

👍 for unit tests

Reply:
Because if we have a single chunk then the padding at the end of the chunk doesn't do anything.
If we have two chunks with AAAA as data and paading size of 0, then the reassembled chunk would look like this:
AAAA0AAAA0
This obviously gives a decoding error. if only AAAA0 needs to be decoded, the decoder just ignores the 0

And yes, you're right about the extra padding size. I forgot that.